sql/catalyst/src/main/scala/org/apache/spark/sql/util/ReuseMap.scala

```diff
@@ -0,0 +1,73 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.util
+
+import scala.collection.mutable.Map
+import scala.language.existentials
+
+import org.apache.spark.sql.catalyst.plans.QueryPlan
+import org.apache.spark.sql.types.StructType
+
+/**
+ * Map of canonicalized plans that can be used to find reuse possibilities.
+ *
+ * To avoid costly canonicalization of a plan:
+ * - we use its schema first to check if it can be replaced to a reused one at all
+ * - we insert it into the map of canonicalized plans only when at least 2 have the same schema
+ */
+class ReuseMap[T <: QueryPlan[_]] {
+  // scalastyle:off structural.type
+  private val map = Map[StructType, (T, Map[T2 forSome { type T2 >: T }, T])]()
```

Comment:
If the existence type `T2`, which stands for the canonicalized type of `T`, is hard to read then we can move T2 to the class definition like this:
```
class ReuseMap[T <: QueryPlan[_], T2 >: T] {
  private val map = Map[StructType, (T, Map[T2, T])]()
```
and initialize the `ReuseMap`s as:
```
      val exchanges = new ReuseMap[Exchange, SparkPlan]()
      val subqueries = new ReuseMap[BaseSubqueryExec, SparkPlan]()
```

Reply:
Ok, https://github.com/apache/spark/pull/28885/commits/5df4c53d8f2805699bbcc2eae99380b59fd3ac3d adds the new T2 type param.